### PR TITLE
User custom start end gcode

### PR DESCRIPTION
**Pull request by PPAC37**
A try to implement #580 

* [x] (Edit : Not solve, but, worked around via the use of a JTabbedPane ) Problem with SelectPanel displaying several SelectJTextAreas (The SelectPanel and Select* seem not to want to do what I want... how to tame them?) 
* [x] implementation to use them in MarlinPlotterInferface / PlotterControl ( Thanks to coliss86 )
* [x] (will not be done. a new panel, to avoid too crowded PlotterSettings)UI review for PlotterSettings 
* [x] placeholder/replacement pattern
* [ ] Review Plotter Start/End GCode Panel ( presentation, resize, ... close on save)
~~~
G28;
G0 X{limit_left+10} Y{limit_top-20} ; Park position
M300;
~~~
give in .gcode
~~~
G28;
G0 X-315.0 Y480.0 ; Park position
M300;
~~~
* [ ] placeholder/replacement pattern cleanning
* [x] FunctionSolver.java CI test (for currently used expressions evaluation eg: +, -  )
* [ ] find the right compromise for the notion of start gcode. PlotterControl adds a G28 XY . ( so the start gcode isn't exactly one anymore... ( in fact the G28 XY should only be done if there is no G28 present in the user's start gcode... likewise during SaveGcode ) )

### CODE REVIEW

**Comment by i-make-robots** [comment]
if you drop the requirement for math inside the parameters so that `{limit_left+10}` becomes `{limit_left}` then it's an easy pattern replacement.

**Comment by i-make-robots** [comment]
> UI review for PlotterSettings 

I feel there should be a basic tab, simulation tab (for marlin time estimation), and advanced user tab (with custom gcode stuff)

File: src/main/java/com/marginallyclever/makelangelo/plotter/plotterControls/ChooseConnection.java [hunk 1]
```diff
@@ -32,9 +32,9 @@
 	private static final long serialVersionUID = 4773092967249064165L;
 
 	private final ToggleButtonIcon bConnect = new ToggleButtonIcon(
-			asList("ButtonConnect", "ButtonDisconnect"),
-			asList("/images/connect.png", "/images/disconnect.png"),
-			asList(Color.GREEN, Color.RED));
+			new ToggleButtonIcon.Item("ButtonConnect", "/images/connect.png", Color.GREEN),
+			new ToggleButtonIcon.Item("ButtonDisconnect", "/images/disconnect.png", Color.RED)
+		);
```
Comment (i-make-robots) [inline src/main/java/com/marginallyclever/makelangelo/plotter/plotterControls/ChooseConnection.java:37]: nice.

File: src/main/java/com/marginallyclever/makelangelo/plotter/settings/PlotterSettingsUserGcode.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		return res.toString();
+	}
+
+	void saveUserGcode(Preferences thisMachineNode) {
+		thisMachineNode.put(USER_GENERAL_START_GCODE, userGeneralStartGcode);
+		thisMachineNode.put(USER_GENERAL_END_GCODE, userGeneralEndGcode);
+		thisMachineNode.put(USER_TOOL_CHANGE_START_GCODE, userToolChangeStartGcode);
+		thisMachineNode.put(USER_TOOL_CHANGE_END_GCODE, userToolChangeEndGcode);
+	}
+	public static final String USER_TOOL_CHANGE_END_GCODE = "userToolChangeEndGcode";
```
Comment (i-make-robots) [inline src/main/java/com/marginallyclever/makelangelo/plotter/settings/PlotterSettingsUserGcode.java:165]: move to top of class?

File: src/main/java/com/marginallyclever/makelangelo/plotter/plotterControls/ChooseConnection.java (hunk 1 above)
Comment (PPAC37) [reply src/main/java/com/marginallyclever/makelangelo/plotter/plotterControls/ChooseConnection.java]: I can only say the same, @coliss86  did a good job. :)

File: src/main/java/com/marginallyclever/makelangelo/plotter/settings/PlotterSettingsUserGcode.java (hunk 2 above)
Comment (PPAC37) [reply src/main/java/com/marginallyclever/makelangelo/plotter/settings/PlotterSettingsUserGcode.java]: Yes, sorry, normally it is done now.

File: src/main/java/com/marginallyclever/makelangelo/plotter/settings/PlotterSettingsUserGcodePanel.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	private final SelectTextArea userGeneralEndGcode;
+	private final SelectTextArea userToolChangeStartGcode;
+	private final SelectTextArea userToolChangeEndGcode;
+
+	public PlotterSettingsUserGcodePanel(Plotter robot) {
+		super();
+		this.myPlotter = robot;
+
+		PlotterSettings settings = myPlotter.getSettings();
+
```
[the remark this answers is not in the document]
Comment (PPAC37) [reply src/main/java/com/marginallyclever/makelangelo/plotter/settings/PlotterSettingsUserGcodePanel.java:34]: I added a SelectTextArea (which I still have to make non-modifiable) to give the list.
But I still have some sorting to do in it as I'm basing it on preferences, there are entries that aren't really useful for start and end gcode...

**Comment by i-make-robots** [comment]
> G0 X{limit_left+10} Y{limit_top-20} ; Park position

Marlin has workspaces (G54-G59?).  Could these be used to store the park position between sessions?

**Comment by PPAC37** [comment]
This PR is no longer relevant.

Thank you coliss86 for taking over and bringing to fruition in #656 the idea that I had trouble finalizing.

**Comment by coliss86** [comment]
Thanks for writting it ;-) 

One thing is still missing is the `FunctionSolver`, I'll look at it in the next days

**Pull request closed without merging**